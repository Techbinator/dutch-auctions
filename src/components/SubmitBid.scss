.price {
  input {
    font-size: 1.7em;
    text-align: right;
  }

  button:disabled,
  button[disabled] {
    background: #ccc;
  }
}
