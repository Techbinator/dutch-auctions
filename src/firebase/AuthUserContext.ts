import * as React from "react";
import { User } from "firebase";
const defaultData: User | null = null;
export const AuthUserContext = React.createContext(defaultData);
