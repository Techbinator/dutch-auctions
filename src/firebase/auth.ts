import { auth } from "./config";

// Sign Up
export const createUserWithEmailAndPassword = (
  email: string,
  password: string
) => auth.createUserWithEmailAndPassword(email, password);

// Sign In
export const signInWithEmailAndPassword = (email: string, password: string) =>
  auth.signInWithEmailAndPassword(email, password);

// Sign out
export const signOut = () => auth.signOut();
