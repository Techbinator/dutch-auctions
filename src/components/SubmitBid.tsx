import * as React from "react";
import { db } from "../firebase";
import Button from "../components/Button";

import "./SubmitBid.scss";
import { IAuction } from "../types/auction.type";
import { calculatePriceAndStatus } from "../helpers/auction";

interface SubmitBidProps {
  userId: string;
  userEmail?: string | null;
  auction: IAuction;
  auctionEnded: boolean;
}

interface ISubmitBidState {
  amount: number;
  loading: boolean;
  error: string;
}

const INITIAL_STATE = {
  amount: 0,
  loading: false,
  error: ""
};

export default class SubmitBid extends React.Component<
  SubmitBidProps,
  ISubmitBidState
> {
  state = INITIAL_STATE;

  onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    //make typescript happy
    switch (event.target.name) {
      case "amount":
        this.setState({ amount: parseFloat(event.target.value) });
        break;
    }
  };
  onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    this.setState({ loading: true });
    const amount = this.state.amount ? this.state.amount : 1;
    const { userId, userEmail, auction } = this.props;
    const auctionId = auction.id;
    const data = { amount, userId, userEmail, auctionId };
    db.submitBid(data)
      .then(() => {
        const auctionBidData = { amount, auctionId };
        db.submitBidToAuction(auctionBidData)
          .then(() => {
            this.setState({ ...INITIAL_STATE, amount });
          })
          .catch((error: Error) => {
            this.setState({ error: error.message, loading: false });
          });
      })
      .catch((error: Error) => {
        this.setState({ error: error.message, loading: false });
      });
  };

  isSameUser = () => {
    return this.props.auction.ownerId === this.props.userId;
  };

  public render() {
    const {
      auction: { currentMaxBid },
      auctionEnded
    } = this.props;
    const priceAndStatus = calculatePriceAndStatus(this.props.auction);

    const { error, loading, amount } = this.state;
    const disableInput =
      this.isSameUser() || priceAndStatus.ended || auctionEnded;

    return (
      <form onSubmit={this.onSubmit} className="price">
        {error && <p className="error">{error}</p>}
        <input
          disabled={disableInput}
          type="number"
          name="amount"
          required
          value={amount || currentMaxBid || 1}
          onChange={this.onChange}
          min={currentMaxBid}
          max={priceAndStatus.price}
          step="0.01"
          maxLength={9}
        />
        <Button
          disabled={disableInput}
          type="primary"
          text={loading ? "Loading..." : "BID"}
        />
      </form>
    );
  }
}
