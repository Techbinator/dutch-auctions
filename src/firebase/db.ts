import { db } from "./config";
import { User, firestore } from "firebase";

interface IGetAuctions {
  authUser: User;
  currentUserAuctions: boolean;
}
export const getAuctions = ({
  authUser,
  currentUserAuctions
}: IGetAuctions) => {
  let auctions = db.collection("auctions").orderBy("endDate", "desc");
  console.log("here");
  if (authUser && currentUserAuctions) {
    auctions = auctions.where(
      "ownerId",
      "==",
      authUser.uid
    ) as firestore.CollectionReference;
  }

  return auctions;
};

interface ICreateNewAuction {
  title: string;
  startingBid: number;
  ownerId: string;
}

export const createNewAuction = (data: ICreateNewAuction) => {
  var startDate = new Date();
  var endDate = new Date();
  endDate.setMinutes(endDate.getMinutes() + 5);
  const auction = {
    title: data.title,
    startDate: startDate.getTime(),
    endDate: endDate.getTime(),
    startingBid: data.startingBid,
    ownerId: data.ownerId,
    currentMaxBid: 0,
    maxBidder: {
      email: "",
      Id: ""
    }
  };
  return db.collection("auctions").add(auction);
};
