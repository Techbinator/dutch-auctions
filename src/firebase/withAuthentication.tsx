import * as React from "react";
import { firebase } from "./";
import { AuthUserContext } from "./AuthUserContext";
import { User } from "firebase";

interface InterfaceState {
  authUser?: User | null;
}

export const withAuthentication = (Component: any) => {
  class WithAuthentication extends React.Component<{}, InterfaceState> {
    state = {
      authUser: null
    };

    public componentDidMount() {
      firebase.auth.onAuthStateChanged(authUser => {
        authUser
          ? this.setState(() => ({ authUser }))
          : this.setState(() => ({ authUser: null }));
      });
    }

    public render() {
      const { authUser } = this.state;

      return (
        <AuthUserContext.Provider value={authUser}>
          <Component authUser={authUser} />
        </AuthUserContext.Provider>
      );
    }
  }
  return WithAuthentication;
};
