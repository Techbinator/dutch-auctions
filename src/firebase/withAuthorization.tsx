import * as React from "react";
import { withRouter } from "react-router-dom";
import { firebase } from "./";
import Loading from "../components/Loading";
import { AuthUserContext } from "./AuthUserContext";

//verify if user is loged in or not
export const withAuthorization = (checkIsLogedin: boolean) => (
  Component: any
) => {
  class WithAuthorization extends React.Component<any, {}> {
    state = {
      loading: true
    };
    _isMounted: boolean = false;
    public componentDidMount() {
      this._isMounted = true;
      const { history } = this.props;
      firebase.auth.onAuthStateChanged(authUser => {
        if (checkIsLogedin && !authUser) {
          history.push("/login");
        } else if (!checkIsLogedin && authUser) {
          history.push("/my_auctions");
        }
        if (this._isMounted) {
          this.setState({ loading: false });
        }
      });
    }

    public componentWillUnmount() {
      this._isMounted = false;
    }

    public render() {
      const { loading } = this.state;
      if (loading) {
        return <Loading />;
      }

      return (
        <AuthUserContext.Consumer>
          {authUser => <Component authUser={authUser} {...this.props} />}
        </AuthUserContext.Consumer>
      );
    }
  }

  return withRouter(WithAuthorization);
};
