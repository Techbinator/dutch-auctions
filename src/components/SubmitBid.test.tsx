import React from "react";
import SubmitBid from "./SubmitBid";
import { shallow } from "enzyme";

import { db } from "../firebase";
var auctions = require("../helpers/__mocks__/auctions.json");
const auction = auctions[0];

const event = {
  preventDefault: jest.fn()
} as any;

describe("SubmitBid component", () => {
  let wrapper: any;
  beforeEach(() => {
    wrapper = shallow(
      <SubmitBid
        userId="useridtest"
        userEmail="useremailtest"
        auction={auction}
        auctionEnded={false}
      />
    );
  });
  it("should render snapshot with default values", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should return that the owner and current user are different", () => {
    expect(wrapper.instance().isSameUser()).toBeFalsy();
  });

  it("should trigger bid submit with the right parameters", () => {
    const submitBidSpy = jest.spyOn(db, "submitBid");
    const startingBidInput = wrapper.find('input[name="amount"]');

    startingBidInput.simulate("change", {
      target: { value: 33, name: "amount" }
    });
    expect(wrapper.state("amount")).toBe(33);

    wrapper.instance().onSubmit(event);
    expect(submitBidSpy).toBeCalledWith({
      amount: 33,
      auctionId: "37gc2mvSPYSXfSqTNlE34qL8Qusssse2",
      userEmail: "useremailtest",
      userId: "useridtest"
    });
  });
});
