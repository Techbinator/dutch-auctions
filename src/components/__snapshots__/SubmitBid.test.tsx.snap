// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitBid component should render snapshot with default values 1`] = `
<form
  className="price"
  onSubmit={[Function]}
>
  <input
    disabled={true}
    max={7.53}
    maxLength={9}
    min={7.53}
    name="amount"
    onChange={[Function]}
    required={true}
    step="0.01"
    type="number"
    value={7.53}
  />
  <_default
    disabled={true}
    text="BID"
    type="primary"
  />
</form>
`;
